### DISCUSSION TITLE
Skip non-blocking retry with DL

### DISCUSSION THREAD

**Discussion by jekanik** [spring-projects/spring-kafka#2054] [Q&A] [answered]
Hi. Now I'm checking retry mechanism and come up with a question. Is there any **legal** way to have "seek to current" + dead-letter w/o non-blocking retry? 
As I see it's possible to set `org.springframework.kafka.annotation.RetryableTopic#timeout` as zero and `org.springframework.kafka.retrytopic.DefaultDestinationTopicResolver#resolveDestinationTopic` will return only DL topic.
But it looks not really convenient. 
Could you pls advice how to implement such approach?

Thanks

**UPD**.: as I see in such case there is no need to use retry logic at all because for DefaultErrorHandler no reties specified. Just set custom commonErrorHandler to ConcurrentKafkaListenerContainerFactory. 
Then for now it's not clear if it's possible to create CommonErrorHandler as bean w/o creating additional bean with ConcurrentKafkaListenerContainerFactory. 
We are using 2.5.11 on prod and such setup is just working fine. Would be nice to know if it's possible to use similar one with 2.8.0.

![image](https://user-images.githubusercontent.com/13684926/146770994-de6215cc-6623-48b6-8ef1-633b7e26cb65.png)

**Comment by garyrussell** [comment, marked as answer]
If you are using Spring Boot; it will auto-wire an error handler (and message converter) bean into the the auto-configured container factory.

If you are not using Spring Boot, you have to do it yourself.

**Reply by jekanik** [reply]
@garyrussell hi one more time :)
I've revisited injection code little bit and run some test on my prod services.
I found out that only `org.springframework.kafka.listener.ErrorHandler` and `org.springframework.kafka.listener.BatchErrorHandler` can be injected from KafkaAnnotationDrivenConfiguration to ConcurrentKafkaListenerContainerFactoryConfigurer.
So even with injected bean for CommonErrorHandler we still get default one in `org.springframework.kafka.listener.KafkaMessageListenerContainer.ListenerConsumer#determineCommonErrorHandler`.

Pls correct me if I'm wrong. 
Thanks.

**Reply by garyrussell** [reply]
Right - if you explicitly provide a legacy error handler, it will be used (in the `ErrorHandlerAdapter`), unless a `CommonErrorHandler` is injected.

Boot 2.6 was changed to allow auto-injection of a `CommonErrorHandler` bean.

https://github.com/spring-projects/spring-boot/pull/27927

If you are using an older Boot with Spring for Apache Kafka 2.8, you will have to inject the common error handler yourself.
